android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java

```diff
@@ -61,116 +108,127 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
-                getInstantMessageAddressesFromContentResolver(id, cr) : null);
-            collections.put("urlAddresses", fieldsSet.contains("urlAddresses") ?
-                getUrlAddressesFromContentResolver(id, cr) : null);
-            collections.put("relationships", fieldsSet.contains("relationships") ?
-                getRelationshipsFromContentResolver(id, cr) : null);
-
-            for (String fieldName : collections.keySet()) {
-              WritableArray value = collections.get(fieldName);
-              if (value != null && value.size() > 0) {
-                contact.putArray(fieldName, value);
-              }
-            }
+      if (fieldsSet.contains("namePrefix")) {
+        FULL_PROJECTION.add(CommonDataKinds.StructuredName.PREFIX);
+      }
 
-            if (note != null && !note.isEmpty()) {
-              contact.putString("note", note);
-            }
-            if (company != null && !company.isEmpty()) {
-              contact.putString("company", company);
-            }
-            if (jobTitle != null && !jobTitle.isEmpty()) {
-              contact.putString("jobTitle", jobTitle);
-            }
+      if (fieldsSet.contains("nameSuffix")) {
+        FULL_PROJECTION.add(CommonDataKinds.StructuredName.SUFFIX);
+      }
 
-            if (fetchSingleContact) {
-              promise.resolve(contact);
+      cursor = cr.query(
+          ContactsContract.Data.CONTENT_URI,
+          FULL_PROJECTION.toArray(new String[FULL_PROJECTION.size()]),
+          selection,
+          selectionArgs.toArray(new String[selectionArgs.size()]),
+          null);
+    }
+    if (cursor != null) {
+      try {
+        // populate list of contacts from cursor query
+        contacts = loadContactsFrom(cursor, fieldsSet);
+
+        WritableArray contactsArray = Arguments.createArray();
+
+        // introduce paging at this level to ensure all data elements
+        // are appropriately mapped to contacts from cursor
+        // NOTE: paging provides essentially no performance improvement, recommended primarily for functional reasons
+        int currentIndex;
+        ArrayList<Contact> contactList = new ArrayList<>(contacts.values());
+        int collectionSize = contactList.size();
+        // convert from pojo to react native
+        for (currentIndex = pageOffset; currentIndex < collectionSize; currentIndex++) {
+          Contact contact = contactList.get(currentIndex);
+
+          // if fetching single contact, short circuit and return contact
+          if (fetchSingleContact) {
+            promise.resolve(contact.toMap(fieldsSet));
+            break;
+          } else {
+            if ((currentIndex - pageOffset) >= pageSize) {
               break;
-            } else {
-              contacts.pushMap(contact);
-              currentIndex++;
             }
+            contactsArray.pushMap(contact.toMap(fieldsSet));
           }
         }
 
         int total = cursor.getCount();
```

Comment:
This return a wrong value - for example is 15 when my device has 6 contacts.

Reply:
good catch! unfortunate there isn't a good way to catch this in current test automation

check.